### PR TITLE
 CI: move the musllinux Cirrus job to GHA, optimize other jobs

### PR DESCRIPTION
**Pull request by andyfaff**
Cirrus is going to start limiting free usage. This PR moves a musllinux_x86_64 job from cirrus to GHA. It also reduces the resources given to the linux_aarch64 job.

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks @andyfaff, this seems like a good idea to me.

For others who were looking for info on the Cirrus CI account change: https://cirrus-ci.org/blog/2023/07/17/limiting-free-usage-of-cirrus-ci/

**Comment by tupui** [comment]
What do you think about self hosted runners? (Either for GH or Cirrus). That could be a more manageable alternative to a full self hosted solution.

I suppose we could also raise that to NumFOCUS and ask if they could think about sharing self hosted workers between projects. The range of service they provide is still unclear to me, but I remember discussions around more tooling or cloud credit.

**Comment by rgommers** [comment]
> What do you think about self hosted runners? (Either for GH or Cirrus)

I'd much rather pay Cirrus if needed, self hosted runners are a last resort that take a lot of work to administer and should be avoided if at all possible.

**Comment by andyfaff** [comment]
closing temporarily so I can continue to experiment in my fork.

**Comment by larsoner** [comment]
>> Cirrus is going to start limiting free usage

> https://cirrus-ci.org/blog/2023/07/17/limiting-free-usage-of-cirrus-ci/

Looks like they say:

> Starting September 1st 2023, there will be an upper monthly limit on free usage equal to 40 compute credits

Yikes we are way over! :scream: 

Screenshot from 2023-07-31 12-16-19


Thanks for looking into this @andyfaff

File: LICENSE.txt
Comment (larsoner) [inline LICENSE.txt:910]: Cruft that this was added?

Comment (andyfaff) [reply LICENSE.txt]: The perils of building scipy and then committing things. I've removed the changes to LICENSE.txt

**Comment by andyfaff** [comment]
40 compute credits is going to be difficult to reach. Whilst the changes here will bring down our usage, it's only going to be scratching the surface.

Things we can do to reduce further:

- make use of compiler caches.
- only make 'nightly' wheels (`linux_aarch64` and `macosx_arm64`) every two weeks instead of every week (or some other reduced frequency).
- Don't regularly test `linux_aarch64` or `macosx_arm64`, we currently do this on every run. Instead rely on the 'nightly' wheel build failing to tell us something is wrong. Alternatively, don't test the wheels when we make them (there's a wheel for each Python version), but perform regular testing on one target.
- For regular tests only perform a reduced set of tests, e.g. `scipy.linalg` would make sense.
- Reduce CPU usage in all the runs. For example, the linux_aarch64 job seems to use the 4-core CPU well during the build phase, but we have underutilisation during test. This PR reduces to 2-core CPU. The same argument applies to building wheels.
Screenshot 2023-08-01 at 08 01 53

**Comment by andyfaff** [comment]
@rgommers I have a question for caching things from a meson build. Are we better off creating and using a ccache, or would it be preferable to cache the meson `build` directory?

**Comment by andyfaff** [comment]
@larsoner,
I think we're good to go now.

- moved musllinux_x86_64 to GHA, Only have linux_aarch64 and macosx_arm64 on regular CI now.
- enabled caching of pip for both of those
- enabled ccache for both of those. A cache hit on linux_aarch64 can reduce the run time by a factor of two.
- reduced resources used by linux_aarch64 from CPU=4 to CPU=2.
- use `-j` option to build/test in parallel.

**Comment by andyfaff** [comment]
I calculate that this PR will quarter our regular CI runs.

Before:
8 CPU * 22 mins (musl) + 19*4 (linux_aarch64) + 10*4 (macosx_arm64) = 292 CPU mins

After:
11*2 + 13*4 = 74 CPU mins

**Comment by andyfaff** [comment]
Using 1 CPU on the linux_aarch64 job decreases the time from 11 * 2 = 22 CPU minutes to ~17 CPU minutes (if there is caching)

**Comment by rgommers** [comment]
> @rgommers I have a question for caching things from a meson build. Are we better off creating and using a ccache, or would it be preferable to cache the meson `build` directory?

Using `ccache`. The comparison between source and build directories relies on time stamps, so restoring that isn't reliable IIRC.

**Comment by rgommers** [comment]
This looks great, thanks @andyfaff!

> I calculate that this PR will quarter our regular CI runs. [...]

Why is the time for macOS arm64 going up from 10 to 13 minutes? Those minutes are far more expensive than the Linux ones, so if it's not random run to run variation, maybe it requires another tweak?

**Comment by tupui** [comment]
> 74 CPU mins

Does that mean that we have a total budget of 10000/74 = 135 runs? That would mean 4 per day.

If that's the case, then I would just use the service for nightly jobs.

**Comment by tupui** [comment]
Another alternative I already propose is to rethink how we run CI and request a maintainer to ask for it to run. Then we can be mindful about what we do and tbh the DX is not that impacted as maintainers reviewing would either way spend time looking at a PR.

I really doubt that GH is going to be free forever as everyone depends more and more on them alone.

**Comment by andyfaff** [comment]
> Why is the time for macOS arm64 going up from 10 to 13 minutes? Those minutes are far more expensive than the Linux ones, so if it's not random run to run variation, maybe it requires another tweak?

Paradoxically caching things didn't help. I don't know why. There doesn't seem to be an ability to control how many CPUs we get.

> Does that mean that we have a total budget of 10000/74 = 135 runs? That would mean 4 per day.

The value of 10000 was Linux minutes. I don't know how the macOS minutes are calculated. But that might be ballpark correct.
It gets worse, that's not including the wheel builds, which chew though resources like there's no tomorrow. That's basically quadrupling the job size - hence the suggestion of dropping to fortnightly.

I don't know whether it's possible to start jobs manually.

> One suggestion: it'd be useful to separate out the "install dependencies", "build" and "test" parts of these jobs in separate steps. That makes the logs much easier to navigate, and where there's a failure it's also useful to see which of those steps failed.

Both these tasks were done in a single test_script, it's only the musl script that was split into multiple scripts.

**Comment by tupui** [comment]
> The value of 10000 was Linux minutes. I don't know how the macOS minutes are calculated. But that might be ballpark correct.
It gets worse, that's not including the wheel builds, which chew though resources like there's no tomorrow. That's basically quadrupling the job size - hence the suggestion of dropping to fortnightly.

Ouch! Then I would suggest to first lay down the plan of what we want to do before tweaking this at all.

**Comment by rgommers** [comment]
> I don't know how the macOS minutes are calculated.

It's now 5x more credits per CPU minute, and after the announced price reductions it's ~3x more credits. And yes, it does not seem possible to request fewer than 4 cores, I think because they have technical constraints that force them to allocate only whole M1 CPUs.

>  Both these tasks were done in a single test_script, it's only the musl script that was split into multiple scripts.

Ah, there is no way to split that `test_script` block out into multiple steps, like on GHA?

**Comment by andyfaff** [comment]
> Ouch! Then I would suggest to first lay down the plan of what we want to do before tweaking this at all.

I think we would to merge this PR even if there weren't constraints on minutes, then go from there.

> Ah, there is no way to split that test_script block out into multiple steps, like on GHA?

There is, I just hadn't done it previously.

**Comment by andyfaff** [comment]
There is a way to do a manual trigger, https://cirrus-ci.org/guide/writing-tasks/#manual-tasks. Ill have to do a little experimenting with it. Probably in another PR though.

**Comment by andyfaff** [comment]
The cirrus manual trigger can be added to a cirrus task.
I don't think manual means that you can click on the task at a random point in time and the task will start.

Think of it like a normal CI run, but with that task in a paused state. The PR page on GH reports a 'neutral' check when the task is paused. You either have to go to the cirrus page for the build to unpause and run the task, or you can do it in a couple of clicks from the checks tab on the PR. It's unclear to me whether any compute credits are used for tasks in a paused state, or how long the paused nature remains.

Perhaps we could manually trigger both the `linux_aarch64` and `macosx_arm64` tasks and only run them when we think it's necessary?

image

image

**Comment by andyfaff** [comment]
Given that the feedback is positive I'm going to merge tomorrow, unless anyone has any objections.

File: ci/cirrus_general_ci.yml
Comment (rgommers) [inline ci/cirrus_general_ci.yml:140]: Is the `-j3` a leftover here from before splitting build/test? Either leaving it out or using `-j4` should be faster. This is `ninja`'s parallelism, which scales way better than `pytest-xdist`.

File: ci/cirrus_general_ci.yml
Comment (rgommers) [inline ci/cirrus_general_ci.yml:147]: This is somehow super slow. On this PR:
```
=== 42915 passed, 2476 skipped, 146 xfailed, 13 xpassed in 567.75s (0:09:27) ===
```
vs. a recent run on another PR:
```
= 42915 passed, 2476 skipped, 12693 deselected, 146 xfailed, 13 xpassed in 314.47s (0:05:14) =
```
Not sure if it's a flaky run or if this is really that much worse. Maybe lower to `-j2`?

**Review by rgommers — APPROVED** [review]
Thanks @andyfaff. This looks ready to go to me, modulo probably a tweak to the parallelism levels on the macOS arm64 job. 

I agree it's a good idea to get this in as is. Regarding the follow-up: I'd say let's see where we are at with usage levels after this for the next week or so, and then reassess. I think starting to pay for some Cirrus CI credits will be a good idea - much easier than other options for the arm64/aarch64 jobs.

File: ci/cirrus_general_ci.yml
Comment (andyfaff) [reply ci/cirrus_general_ci.yml:140]: I can take it out. I was trying to optimize build/test usage of CPU, and doing this by experimentation rather than knowing things.

File: ci/cirrus_general_ci.yml
Comment (andyfaff) [reply ci/cirrus_general_ci.yml:147]: I do think there is some variability, perhaps I shouldn't install pytest-xdist or use the j flag?

Comment (andyfaff) [reply ci/cirrus_general_ci.yml]: I give up. I look at the CPU usage graph, and it looks way busier with j3 compared to no j flag, yet the j3 is slower. I wonder if parallel testing just results in a whole heap of thrashing.

Comment (rgommers) [reply ci/cirrus_general_ci.yml]: It must do, yes. I agree that it's frustrating and hard to make sense of. Let's merge this as it is now?

**Pull request merged**
